core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1293,12 +1269,9 @@ class Log(@volatile var dir: File,
   private def maybeRoll(messagesSize: Int, maxTimestampInMessages: Long, maxOffsetInMessages: Long): LogSegment = {
     val segment = activeSegment
     val now = time.milliseconds
-    val reachedRollMs = segment.timeWaitedForRoll(now, maxTimestampInMessages) > config.segmentMs - segment.rollJitterMs
-    if (segment.size > config.segmentSize - messagesSize ||
-        (segment.size > 0 && reachedRollMs) ||
-        segment.index.isFull || segment.timeIndex.isFull || !segment.canConvertToRelativeOffset(maxOffsetInMessages)) {
+    if (segment.shouldRoll(messagesSize, maxTimestampInMessages, maxOffsetInMessages, now)) {
```

Comment:
@ijuma This code used to check `config.segmentSize` that was dynamic (could be updated at topic-level). Now it checks a value that is cached in the log segment and is not updated. Was that change deliberate? I was testing dynamic update of broker defaults by checking for log rolling, but I rebased today and the test no longer works. I think the failure is due to this change. If it was deliberate, I will change the test, but wanted to check first. Thanks.

Reply:
I just followed the same pattern as `rollJitterMs` and `indexIntervalBytes` and didn't consider the dynamic update case. Do we want to be able to update it dynamically for an existing segment? Or does it make more sense to do it for new segments? Not taking a position either way, just asking.

Reply:
I thought the others were just missed out and was going to fix them. But I don't have an opinion either way as well, so will just leave it with your update and fix the test. Thanks.

Reply:
@junrao, what do you think, would it be useful to be able to dynamically update the max segment size and max segment ms for a segment that is currently open?

Reply:
Yes, semantic wise, it does seem the dynamic config change should take effect on the existing segments as well.